### PR TITLE
SILOptimizer: fix non-deterministic behavior in RedundantLoadElimination and DeadStoreElimination.

### PR DESCRIPTION
**Pull request by eeckstein**
Replace some DenseSets, which are used for iteration, with vectors.
This also needed some rewrite in LSLocation::reduce() to rule out quadratic complexity.

SR-8844
rdar://problem/44762620

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: lib/SILOptimizer/Utils/LoadStoreOptUtils.cpp [hunk 1]
```diff
@@ -185,36 +185,67 @@ LSLocation::expand(LSLocation Base, SILModule *M, LSLocationList &Locs,
   }
 }
 
-bool
-LSLocation::reduce(LSLocation Base, SILModule *M, LSLocationSet &Locs) {
+/// Does a location list contain a specific location?
+static bool contains(const LSLocationList &Locs, const LSLocation &L) {
```
Comment (jrose-apple) [inline lib/SILOptimizer/Utils/LoadStoreOptUtils.cpp]: `llvm::is_contained` already does this.

Comment (eeckstein) [reply lib/SILOptimizer/Utils/LoadStoreOptUtils.cpp]: Ah, thanks. I'll change it

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**